utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java

```diff
@@ -0,0 +1,497 @@
+
+    /**
+     * Invoked from within {@link #processEventQueue()} when we can't process the queue for some reason. This is likely
+     * caused by a downstream subscriber throwing an exception from {@code onNext}, which it should never do.
+     *
+     * <p>Here we try our best to fail all of the entries in the queue, so that no callers have "stuck" futures.
+     */
+    private void panicAndDie(Throwable cause) {
+        try {
+            // Create exception here instead of in supplier to preserve a more-useful stack trace.
+            RuntimeException failure = new IllegalStateException("Encountered fatal error in publisher", cause);
+            rejectException.compareAndSet(null, () -> failure);
+            entryTypesToFail.addAll(asList(QueueEntry.Type.values()));
+            subscriber.onError(failure);
+
+            while (true) {
+                QueueEntry<T> entry = eventQueue.poll();
+                if (entry == null) {
+                    break;
+                }
+                entry.resultFuture.completeExceptionally(failure);
+            }
+        } catch (Throwable t) {
+            t.addSuppressed(cause);
+            log.error(() -> "Failed while processing a failure. This could result in stuck futures.", t);
+        }
+    }
+
+    /**
+     * Ensure that {@link #rejectException} is null. If it is not, throw the exception.
+     */
+    private void checkRejectException() {
+        if (rejectException.get() != null) {
+            throw rejectException.get().get();
+        }
+    }
+
+    /**
+     * Set the {@link #rejectException}, if it is null. If it is not, throw the exception.
+     */
+    private void setRejectExceptionOrThrow(Supplier<RuntimeException> rejectedException) {
+        if (!rejectException.compareAndSet(null, rejectedException)) {
+            throw rejectException.get().get();
+        }
+    }
+
+    /**
+     * The subscription passed to the first {@link #subscriber} that subscribes to this publisher. This allows the downstream
+     * subscriber to request for more {@code onNext} calls or to {@code cancel} the stream of messages.
+     */
+    private class SubscriptionImpl implements Subscription {
+        @Override
+        public void request(long n) {
+            log.trace(() -> "Received request() with " + n);
+            if (n <= 0) {
+                // Create exception here instead of in supplier to preserve a more-useful stack trace.
+                IllegalArgumentException failure = new IllegalArgumentException("A downstream publisher requested an invalid "
+                                                                                + "amount of data: " + n);
+                rejectException.compareAndSet(null, () -> failure);
+                eventQueue.add(new OnErrorQueueEntry<>(failure));
+                entryTypesToFail.add(ON_NEXT);
+                processEventQueue();
+            } else {
+                long newDemand = outstandingDemand.addAndGet(n);
```

Comment:
> A Subscription MUST support an unbounded number of calls to request and MUST support a demand up to 2^63-1 (java.lang.Long.MAX_VALUE). 

I read this "up to" as "but not greater than".

> A demand equal or greater than 2^63-1 (java.lang.Long.MAX_VALUE) MAY be considered by the Publisher as “effectively unbounded”.

Since this said "may" I was being lazy, but you're probably late - I shouldn't be lazy.